README.md
# Bert

Bert is a very simple download manager written for the Raspberry Pi (or any other embedded Linux device). 

![Screenshot of the web interface](/screenshot.png?raw=true "Screenshot of the web interface currently downloading the latest Kernel version")

Use case: I often need to download large files which I don't want to download via my computer since it bothers me having a process which can't interrupted by accident or on purpose. The Raspberry Pi is near the router and can download all the day without disturbing anybody or anything.

Bert uses `wget` and some other linux command line tools (e.g. `du`, `df`) as backend. NodeJS provides the server and application logic and invokes `wget` through some wrapper object and parses it output. 

[Click here to read more about Bert on my blog!](http://maxstrauch.github.io/2016/01/08/bert-downloads-it.html)

# Features

 - Simple
 - The web interface proivdes an easy way to add or monitor downloads
 - Queue behaviour: only one download active download at a time
 - Run user script from web interface (very useful to copy the downloaded files from the `downloads/` folder to an USB thumb drive without `ssh` into the Pi)
 - Clear the downloads folder from web interface
 - Shutdown server from web interface
 - Detection of downloading the same file multiple times

# Run it!

You can run it by simply invoking

	node server.js

I have created a `cron` entry in my Pi with `crontab -e` by adding the line `@reboot /bin/bash --login /home/pi/script.sh > /home/pi/work.log` which will excute a special script on reboot. `/home/pi/script.sh` contains the following:

	#!/bin/sh
	sleep 120
	(/usr/local/bin/node server.js > out.log) &

I don't know why there is a `sleep 120` anymore. But you all know the rule: never change a running system!

The configuration is done in the first 32 lines of `server.js` - each parameter is documented there.

# License

Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0). Click [here](http://creativecommons.org/licenses/by-sa/4.0/) for details. The license applies to the entire source code.

`This program is distributed WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.`

# Trivia

I'm a huge fan of NCIS. Guess who Bert is ... 

![Bert](/www/logo.png?raw=true)



CHANGELOG.txt
Bert 0.42.1, 2016-01-04
---------------------
 - Added automatic download folder creation if not present
 - Tweaked client script to return false on link click and
   therefore don't show the hash character in the URL
 - Added this `CHANGELOG.txt` file to project
 - Added a favicon to the index.html page (and also a new
   MIME type to serve *.ico files)
 - Added a screenshot to the description
 - Removed some debug code `from www/script.js`

Bert 0.42, 2016-01-04
---------------------
 - Initial version

www/script.js
// Converts seconds to a more convenient format
// @see: http://stackoverflow.com/a/13368349/2429611
function toHHMMSS(n) {
    var seconds = Math.floor(n),
        hours = Math.floor(seconds / 3600);
    seconds -= hours*3600;
    var minutes = Math.floor(seconds / 60);
    seconds -= minutes*60;

    if (hours   < 10) {hours   = "0"+hours;}
    if (minutes < 10) {minutes = "0"+minutes;}
    if (seconds < 10) {seconds = "0"+seconds;}
    return hours+':'+minutes+':'+seconds;
}

// Format bytes human read-able
// @see: http://stackoverflow.com/a/18650828/2429611
function bytesToSize(bytes) {
   var sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
   if (bytes == 0) return '0 Byte';
   var i = parseInt(Math.floor(Math.log(bytes) / Math.log(1024)));
   return Math.round(bytes / Math.pow(1024, i), 2) + ' ' + sizes[i];
}

// Shortcut to get a DOM element by it's id
function $(id) {
  return document.getElementById(id);
}

// Shortcut to make a new DOM element and apply
// different attributes to it
function $make(tag, opts) {
  var elem = document.createElement(tag);

  if (opts && opts.className) {
    elem.className = opts.className;
  }

  if (opts && opts.attr) {
    for (var attr in opts.attr) {
      elem.setAttribute(attr, opts.attr[attr]);
    }
  }

  if (opts && opts.id) {
    elem.id = opts.id;
  }

  if (opts && opts.html) {
    elem.innerHTML = opts.html;
  }

  if (opts && opts.onclick) {
    elem.onclick = opts.onclick;
  }

  if (opts && opts.parent) {
    opts.parent.appendChild(elem);
  }

  return elem;
}

// Simple AJAX function to communicate with the backend
function $ajax(url, cb, get) {
  // Crate the url (with GET parameters if supplied)
  var totalURL = url;
  var params = '?';
  for (var key in get) {
    var value = encodeURIComponent(get[key]).replace(/%20/g,'+');
    params += key + '=' + value + '&';
  }
  if (params.length == 1) {
    params = '';
  } else {
    params = params.substring(0, params.length - 1);
  }
  totalURL = url + params;

  // Run the request
  var xmlhttp;
  if (window.XMLHttpRequest)  {
      xmlhttp=new XMLHttpRequest();
  } else {
      xmlhttp=new ActiveXObject("Microsoft.XMLHTTP");
  }

  xmlhttp.onreadystatechange = function() {
    if (xmlhttp.readyState == 4) {
      if (xmlhttp.status == 200) {
        var data = xmlhttp.responseText;
        if (data.length > 0) {
          cb(JSON.parse(xmlhttp.responseText));
          return;
        }
      }

      // No connection to the backend any more
      cry();
    }
  }       
  xmlhttp.open("GET", totalURL, true);
  xmlhttp.send();
}

// Info board object providing short messages just like
// in OS X
function InfoBoard(id, infoDur, alertDur) {
  var board = $(id);

  var display = function(str, className, dur) {
    // Create the element
    var elem = $make('li', {
      'className': 'mmsg ' + className, 
      'parent': board
    });

    // Append the text
    elem.appendChild(document.createTextNode(str));

    // Delete element after time expired
    window.setTimeout(function() {
      elem.className = elem.className + " fadeOut";
      window.setTimeout(function() {
        elem.parentNode.removeChild(elem);
      }, 420);
    }, dur);
  };

  this.alert = function(str) {
    display(str, "err", alertDur);
  };

  this.info = function(str) {
    display(str, "nfo", infoDur);
  };

  return {
    alert: this.alert,
    info: this.info
  };
}

// Main update function for the list
function update() {
  $ajax('/rest/', function(ret) {      
    if (ret.state != 200) {
      // FIXME: make me nicer
      infoBoard.alert("Can't update the download list (a bug)!");
      return;
    }

    // Sort list by state
    ret.all.sort(function(a, b) {
      if (a.state == 4) {
        return 1;
      }

      if (b.state == 4) {
        return -1;
      }

      if (a.state == 1 || b.state == 1) {
        return (b.state == 1 ? 1 : -1) * 1000;
      }
      
      if (a.state == b.state && b.state == 0) {
        return a.id > b.id ? 1 : -1;          
      }

      if (a.state == 0 || b.state == 0) {
        return (b.state == 0 ? 1 : -1) * 500;
      }
      
      return a.id - b.id;
    });

    // Remove all old entries
    var listNode = $('output');
    while (listNode.firstChild) {
      listNode.removeChild(listNode.firstChild);
    }

    // Create and add new entries
    for (var i = 0; i < ret.all.length; i++) {
      var li = $make('li', {'parent': listNode});
      var e = ret.all[i];

      // Add a progress bar, if downloading
      if (e.state == 1) {
        var prog = $make('div', {
          'className': 'progress',
          'parent': li
        });
        $make('div', {
          'parent': prog
        }).style.width = e.percent + '%';
      }

      // Create the header
      var heading = $make('h1', {'parent': li});
      heading.title = e.name;

      if (e.state == 0) {
        $make('em', {'className': 'icon-hourglass', 'parent': heading});
      } else if (e.state == 1) {
        $make('em', {'className': 'icon-download', 'parent': heading});
      } else if (e.state == 3) {
        $make('em', {'className': 'icon-fail', 'parent': heading});
      } else if (e.state == 4) {
        $make('em', {'className': 'icon-file', 'parent': heading});
      } else {
        if (e.success) {
          $make('em', {'className': 'icon-success', 'parent': heading});
        } else {
          $make('em', {'className': 'icon-fail', 'parent': heading});
        }
      }

      if (e.state == 4) {
        heading.appendChild(document.createTextNode(e.name));
      } else {
        heading.appendChild(document.createTextNode(e.url));
      }
      
      // Add delete action only if not already deleted or downloaded
      if (e.state == 0 || e.state == 1) {
        $make('a', {
          'className': 'action',
          'onclick': function(e) {
            var id = this.getAttribute('data-id');
            $ajax('/rest/', function(ret1) {
              // FIXME: do something useful here
            }, {'cmd': 'stop', 'id': id});
          },
          'attr': {'data-id': e.id},
          'parent': li
        }).href = '#';
      }

      // Render the meta information area
      var div = $make('div', {'className': 'meta', 'parent': li});

      if (e.state == 1) { // Only on download ...
        $make('label', {'html': 'Progress:', 'parent': div});
        $make('strong', {'html': e.percent + '%', 'parent': div});

        $make('label', {'html': 'ETA:', 'parent': div});
        $make('strong', {'html': toHHMMSS(e.eta), 'parent': div});

        $make('label', {'html': 'Speed:', 'parent': div});
        $make('strong', {'html': e.speed, 'parent': div});
      }

      if (e.state == 4) {
        $make('label', {'html': 'Size:', 'parent': div});
        $make('strong', {'html': bytesToSize(e.size), 'parent': div});
      } else {

        if (e.state >= 1 && e.filename.length > 0) {
          $make('label', {'html': 'Filename:', 'parent': div});
          $make('strong', {'html': e.name, 'parent': div});
        } else if (e.state > 1 && e.size > 0) {
          $make('label', {'html': 'Size:', 'parent': div});
          $make('strong', {'html': bytesToSize(e.size), 'parent': div});

          $make('label', {'html': 'Successful?', 'parent': div});
          $make('strong', {'html': e.success ? 'Yes' : 'No', 'parent': div});
        } else if (e.success == false && e.errmsg.length > 0) {
          $make('label', {'html': 'Successful?', 'parent': div});
          $make('strong', {'html': e.success ? 'Yes' : 'No', 'parent': div});

          $make('label', {'html': 'Error message:', 'parent': div});
          $make('strong', {'html': e.errmsg, 'parent': div});
        }
      }
    }
  }, {'cmd': 'list'});
}

// Update system stats (e.g. used disk space)
function updateStats() {
  $ajax('/rest/', function(ret) {
      if (ret.used < 0) {
        $('diskusage').style.width = '0%';
        $('useddsk').innerHTML = 'N/A';
        return;
      }
      $('diskusage').style.width = ret.percent + '%';
      $('useddsk').innerHTML = Math.round(ret.percent*10)/10 + '% (' + bytesToSize(ret.used) + ')';
  }, {'cmd': 'dskstat'});
}

// Check for user script execution and display
// the result
function updateScriptExec() {
  $ajax('/rest/', function(ret) {
      if (ret.state == 200) {
        // Disable controls and show overlay
        $('script-data-out').innerHTML = '';
        $('script-exit').className = 'button disabled';
        $('script-out').style.display = 'block';

        // Get the data from stdout
        $ajax('/rest/', function(ret) {
          // Set it
          $('script-data-out').innerHTML = ret.data;

          // If finished ...
          if (ret.finished) {
            $ajax('/rest/', function(ret) {
              // FIXME: really don't care?
            }, {'cmd': 'user-script', 'exit': true});

            // Enable close button of script overlay
            $('script-exit').className = 'button';
            $('script-exit').onclick = function() {
              this.onclick = undefined;
              $('script-out').style.display = 'none';
              return false;
            };
          }
        }, {'cmd': 'user-script'});
      }
    }, {'cmd': 'user-script', 'test': true}
  );
}

// Reload the entire page
function reload() {
  window.location.href = '/';
}

// When the backend connection is lost execute this function
// and display an error message and show the overlay for this
// case
function cry() {
  infoBoard.alert("Failed to connect to the backend!");
  for (var i = 0; i < intervals.length; i++) {
    clearInterval(intervals[i]);
  }
  $('fail').style.display = 'block';
}

// Download intervals
var intervals = [];

// Message board
var infoBoard;

window.onload = function() {
  infoBoard = new InfoBoard("mmsg-board", 2100, 5000);

  // Attach listener to download field and button
  // and allow for ENTER to start download directly
  // from the text field
  var loadFile = function(e) {
    var opts;
    if ($('force').checked) {
      opts = {'url': $('addr').value, 'force': 1};
    } else {
      opts = {'url': $('addr').value};
    }

    $ajax('/rest/', function(ret) {
      if (ret.state == 200) {
        infoBoard.info("Download added to the list.");
      } else if (ret.state == 501) {
        infoBoard.alert("The same URL was previously added to the downloads list.");
      } else if (ret.state == 500) {
        infoBoard.alert("The given string might not be an URL!");
      } else {
        infoBoard.alert("Couldn't add download to list (error code #" + ret.state + ")!");
      }
    }, opts);
    $('addr').value = '';
    $('force').checked = false;
    return false;
  };

  $("load").onclick = loadFile;
  $("addr").onkeyup = function(e) {
    if (e.keyCode == 13) {
      loadFile(e);
    }
    return false;
  };

  // Handle clear all downloads request
  $('on-clear-all').onclick = function(e) {
    if (confirm('Do you really want to remove all downloads?')) {
      $ajax('/rest/', function(ret) {
          if (ret.state == 200) {
            infoBoard.info("Downloads removed, list cleared.");
          } else {
            infoBoard.alert("Error while removing downloads.");
          }
        }, {'cmd': 'clear-all'}
      );
    }
    return false;
  };

  // Handle user script execution
  $('on-user-script').onclick = function(e) {
    $ajax('/rest/', function(ret) {
        updateScriptExec();
      }, {'cmd': 'user-script', 'run': true}
    );
    return false;
  };

  // Handle shutdown request
  $('on-shutdown').onclick = function(e) {
    var code = prompt('Enter keycode to shutdown the entire server:');
    if (code.length > 0) {
      $ajax('/rest/', function(ret) {
          if (ret.state == 200) {
            infoBoard.info("Server will shutdown now.");
          } else {
            infoBoard.alert("Can't shutdown the server.");
          }
        }, {'cmd': 'halt', 'code': code}
      );
    }

    return false;
  };

  // Create update intervals to keep the download list
  // and stats up-to-date (with different periodicities)
  update();
  intervals.push(setInterval(update, 2500));

  updateStats();
  intervals.push(setInterval(updateStats, 10000));

  updateScriptExec();
  intervals.push(setInterval(updateScriptExec, 2000));

  // Get server version info (only on start up)
  $ajax('/rest/', function(ret) {
    $('versions').innerHTML = ret.app + ' <b>on</b> ' + ret.node;
  }, {'cmd': 'versions'});
}

ScriptRunner.js
var spawn = require('child_process').spawn;
var events = require('events');

// Simple script runner wraper to catch all output of the script
// with the script as argument
function ScriptRunner(scriptName) {
	// State of this object where:
	//   0 = Empty/Reset, 1 = Running, 2 = Finished
	var state = 0;

	// Text received from stdout (and also  stderr)
	var stdout = '';

	// The process handle
	var handle;

	function log(str) {
		console.log(' *** [' + scriptName + '] ' + str)
	}

	// Runs the script attached to this object; should return
	// a boolean whether it works or not
	this.run = function() {
		log('Going to run script')

		// Create new process
		handle = spawn(scriptName);
		state = 1; // Assume that it's running
			
		handle.on('exit', function (code) {
			log('event=exit; EXIT_CODE=' + code)
			state = 2; // Now finished
		});
		
		// Pipe stout and stderr output to the buffer
		handle.stdout.on('data', function (data) {
			stdout += data.toString();
		});

		handle.stderr.on('data', function (data) {
			stdout += data.toString();
		});

		handle.stdin.end();

		// FIXME: test if script is really running
		return true;
	};

	// Clear the object (for a 2nd run)
	this.clear = function() {
		state = 0;
		handle = undefined;
		stdout = '';
		log('Clear triggered')
	};

	// Expose API
	return {
		getState: function() { return state; },
		getOutput: function() { return stdout; },
		run: this.run,
		clear: this.clear
	};
}

exports.ScriptRunner = ScriptRunner;


NodeDownloader.js
var	spawn = require('child_process').spawn,
    events = require('events');

// Creates a new download object with the directory
// where the downloaded file should be saved
function NodeDownloader(dirToSave) {

	// The object of the new process
	var dl;

	// Event emitter
	var eventEmitter = new events.EventEmitter();

	// Last progress value (fire event only on progress change
	// otherwise we would pipe too many calls to the server (maybe))
	var lastProgress;
	
	// General buffer object
	var buffer = '';
	
	// State of this download (see nextLine())
	var state = 0;

	// Other status information variables
	var httpState = -1, retval = -1, contentLength = -1;
	var progress, bytesReceived, downloadRate, eta;
	var saveTo;

	// stopDownload() not yet called
	var wasKilled = false;

	// add the Trim function to javascript
	String.prototype.trim = function() {
		return this.replace(/^\s+|\s+$/g, '');
	}	
	
	// Test if the download was successfull
	this.wasSuccessfull = function() {
		return httpState == 200 && retval < 1;
	}

	// Get the estimated download time outputted by wget
	this.getETA = function() {
		return eta;
	}

	// The name of the downloaded file (from wget containing
	// also the path from the base directory)
	this.getSaveTo = function() {
		return saveTo;
	}

	// Like getSaveTo() but only the filename of the 
	// downloaded file
	this.getName = function() {
		return saveTo.substr(dirToSave.length);
	}
	
	// This function parses every line outputted by wget and
	// saves important bits (e.g. progress) and calls the 
	// listening objects
	var nextLine = function(line) {
		var lline = line.toLowerCase();
		var index;

		switch (state) {
			case 0: // INITIAL
				if (lline.indexOf("response begin") > 0) {
					state = 1;
				} else {
					// Skip lines
				}
				return;

			case 1: // READ_RESPONSE
					if (lline.indexOf("response end") > 0) {
						state = 2;
					} else if (lline.indexOf("http/1.") == 0) {
						httpState = line.match(/[0-9]{3}/);
					} else if ((index = lline.indexOf(": ")) > 0) {
						var key = line.substr(0, index).trim();
						var value = line.substr(index + 1).trim();

						if (key.toLowerCase().indexOf("content-length") != -1) {
							contentLength = value;
						}
					} else {
						// Unknown, skip
					}
				return;

			case 2: // AFTER_RESPONSE
				if (lline.indexOf("response begin") > 0) {
					state = 1;
				} else if ((index = lline.indexOf(dirToSave.toLowerCase())) != -1) {
					line = line.substr(index);
					if ((index = line.indexOf('«')) != -1) {
						saveTo = line.substr(0, index);
					} else {
						saveTo = line.substr(0, line.indexOf('\''));
					}
				} else {
					// Wrap into an try-catch block since sometimes there 
					// occur parsing errors due to missing characters which
					// aren't flushed out of the wget char buffer yet
					try {
						if(line.indexOf("..........") != -1) {
							var regExp = new RegExp('^.*?([0-9a-zA-Z]+).*?[\. ]+.*?([0-9]+)%.*?([a-zA-Z0-9\,\.]+).*?([a-zA-Z0-9\,\.]+)\s*$');
							var prog = line.match(regExp);

							progress = parseInt(prog[2]);
							
							bytesReceived = prog[1];
							var c = bytesReceived.charAt(bytesReceived.length - 1);
							if (c == 'K') {
								bytesReceived = parseInt(bytesReceived.substr(0, bytesReceived.length - 1)) * 1024;
							} else if (c == 'M') {
								bytesReceived = parseInt(bytesReceived.substr(0, bytesReceived.length - 1)) * 1024 * 1024;
							}

							downloadRate = prog[3];
							eta = prog[4];

							var min = eta.match(/([0-9]+)m/);
							if (min != null && min.length > 0) {
								min = parseInt(min[1]);
							} else {
								min = 0;
							}
							var second = eta.match(/([0-9]+)s/);
							if (second != null && second.length > 0) {
								second = parseInt(second[1]);
							} else {
								second = 0;
							}
							eta = min * 60 + second;

							// call only when percentage changed
							if(lastProgress != progress) {
								lastProgress = progress;
								// call the event
								eventEmitter.emit('progress', progress, downloadRate);
							}
						}

					} catch(err) {
						console.log(" *** Downloader error: " + err);
					}
				}
				return;

			default:
				return;
		}
	};

	// That's the master function which starts the download of a file
	// and attaches all necessary listeners
	this.downloadFile = function(file) {
		dl = spawn('wget', ['-d', '-P' + dirToSave, file]);
		
		dl.on('exit', function (code) {
			retval = code;
			if (!wasKilled) { // Don't send finished event iff killed
				eventEmitter.emit('finished');
			}
		});
		
		dl.stderr.on('data', function (data) {
			buffer += data.toString();
			
			var line = '', index = -1;
			for (var i = 0; i < buffer.length; i++) {
				if (buffer.charAt(i) == '\n') {
					nextLine(line);
					line = '';
					index = i;
				} else {
					line += buffer.charAt(i);
				}
			}

			if (index > 0 && index < buffer.length) {
				buffer = buffer.substr(index);
			}
		});
		dl.stdin.end();
	}
	
	// Stops the current download and the wget process which
	// succs the file down
	this.stopDownload = function() {
		wasKilled = true;
		dl.kill();
	}
	
	// Expose API
	return {
		getSaveTo: this.getSaveTo,
		getETA: this.getETA,
		wasSuccessfull: this.wasSuccessfull,
		stopDownload: this.stopDownload,
		downloadFile: this.downloadFile,
		getName: this.getName,
		eventEmitter: eventEmitter
	};
}

exports.NodeDownloader = NodeDownloader;


server.js
// Required modules
var http = require('http');
var url = require('url');
var fs = require('fs');
var downloader = require('./NodeDownloader.js');

// --------------------------------------------------------------------
// Configuration section

// Socket port on which the server listens
var SERVER_PORT = 8080;
// IP address or hostname on which the server listens
var SERVER_HOST = "127.0.0.1";
// Path to the directory containing the downloaded files. This path 
// must end with a slash
var DOWNLOAD_DIR = 'downloads/';
// Name of this application (DO NOT CHANGE!)
var APP_NAME = 'Bert';
// Version of this application (DO NOT CHANGE!)
var APP_VERSION = '0.42.2';
// Script to be executed when the user clicks on "Run user script".
// The second argument takes the script name relative to this file
var SCRIPT = require('./ScriptRunner.js').ScriptRunner('./script.sh');
// The 'password' used to trigger a shutdown command of the
// system the server currently runs on
var SHUTDOWN_CODE = 'abby';
// The shell command executed when a shutdown is triggered
var SHUTDOWN_CMD = 'sleep 10';
// Queue database file name (where the waiting downlaods are saved)
var QUEUE_DB_FN = './queue.json';

// --------------------------------------------------------------------

// Download ID counter (only used internally)
var ID_COUNTER = 0;

// List of all downlaods
var ALL = [];

// The current download (or false if there is no current download)
var CURRENT = false;

// Removes a directory and all its content
// @see https://gist.github.com/liangzan/807712
function rmDir(dirPath) { //: boolean
  try {
  	var files = fs.readdirSync(dirPath);
  } catch(e) { 
  	return false;
  }

  if (files.length > 0) {
    for (var i = 0; i < files.length; i++) {
      var filePath = dirPath + '/' + files[i];
      if (fs.statSync(filePath).isFile()) {
        fs.unlinkSync(filePath);
      } else {
        rmDir(filePath);
      }
    }
  }
  fs.rmdirSync(dirPath);
  return true;
}

// Creates the download directory
function createDownloadDir() {
	// Make sure that there is a downloads directory
	try {
		if (fs.statSync(DOWNLOAD_DIR).isDirectory()) {
			return; // Everything fine
		}
	} catch(e) {
		// Don't watch; (re-)create directory
	}

	fs.mkdir(DOWNLOAD_DIR,function(e){
	    if(e && !(e.code === 'EEXIST')){
	        console.log(' *** Error creating dir `' + DOWNLOAD_DIR + '`: ' + e);
	    } else {
	    	console.log(' *** Oops! There was no `' + DOWNLOAD_DIR + '` dir');
	    }
	});
}

// Calculates the used disk space of the downloads 
// by invoking the `du` command
// FIXME: this function might not work properly (test & correct)
function usedDsk() {
	createDownloadDir();

	var result = -1;
	try {
		var result = parseInt(
			require('child_process')
			.execSync('du -ks ./' + DOWNLOAD_DIR)
			.toString()
		) * 1024;
	} catch(e) {
		result = -1; // Nothing fancy here
	}
	return result;
}

// Calculates the free disk space by parsing the `df` output
// FIXME: this function might not work properly (test & correct)
function totalDsk() {
	createDownloadDir();

	var result = -1, totalDskSpace = -1;
	try {
		totalDskSpace = parseInt(
			require('child_process')
			.execSync('df -Pk ./' + DOWNLOAD_DIR)
			.toString()
			.split(/\n/g)[1]
			.replace(/.*?([0-9]+)(.+$)/i,'$1')
		) * 1024;
	} catch(e) {
		totalDskSpace = -1; // Nothing fancy here
	}
	return totalDskSpace;
}

// Simple function to provide some MIME types needed by the
// server
function mimeLookup(extension) {
	var list = {
		'htm': 'text/html',
		'html': 'text/html',
		'htmls': 'text/html',
		'css': 'text/css',
		'js': 'text/javascript',
		'svg': 'image/svg+xml',
		'eot': 'application/vnd.ms-fontobject',
		'ttf': 'font/truetype',
		'woff': 'application/font-woff',
		'woff2': 'application/font-woff2',
		'png': 'image/png',
		'ico': 'image/x-icon'
	};
	for (var key in list) {
		if (key === extension) {
			return list[key];
		}
	}
	return 'application/octet-stream';
}

// Very basic HTTP error message assembler
function cry(res, statusCode, message) {
	res.writeHead(statusCode, {'Content-Type': 'text/html'});
	res.write('<!DOCTYPE html><html><head><title>Error report</title>');
	res.write('</head><body>');
	res.write('<h1>HTTP ' + statusCode + '</h1>');
	if (message) {
		res.write('<p>' + message + '</p>');
	}
	res.end('</body></html>');
}

// This function is called when a new download should start.
function onStart() {
	if (CURRENT != false || ALL.length < 1) {
		return; // Nothing to do here
	}

	// Get the next download
	CURRENT = false;
	for (var i = 0; i < ALL.length; i++) {
		if (ALL[i].state == 0) {
			CURRENT = ALL[i];
			break;
		}
	}
	if (CURRENT == false) {
		return; // Nothing to do (?!)
	}

	// Begin the download process
	CURRENT.downloader.downloadFile(CURRENT.url);
	CURRENT.state = 1; // Downloading

	// Event listeners
	CURRENT.downloader.eventEmitter.on('progress', function(percent, speed) {
		CURRENT.percent = percent;
		CURRENT.speed = speed;
		CURRENT.eta = CURRENT.downloader.getETA();
		CURRENT.filename = CURRENT.downloader.getSaveTo();
		CURRENT.name = CURRENT.downloader.getName();
	});
	CURRENT.downloader.eventEmitter.on('finished', function() {
		CURRENT.state = 2;
		CURRENT.success = CURRENT.downloader.wasSuccessfull();
		try {
			var stat = fs.statSync(CURRENT.filename);
			CURRENT.size = stat.size;
		} catch (e) {
			CURRENT.size = 0;
			CURRENT.success = false;
			CURRENT.errmsg = e.toString();
		}
		CURRENT = false;
		persist();

		// Let the next download begin (if any)
		onStart();
	});
}

// Persists the current running download and all waiting
// downloads to the backup file (JSON content)
function persist() {
	var output = [];

	// Pull out all waiting downloads
	for (var i = 0; i < ALL.length; i++) {
		if (ALL[i].state == 0 /* queued */ || 
				ALL[i].state == 1 /* active */) {
			output.push(ALL[i].url);
		}
	}

	// Save data to file
	fs.writeFile(QUEUE_DB_FN, JSON.stringify(output), function(err) {
	    if(err) {
	        console.log(' *** Failed to write: ' + QUEUE_DB_FN + ' (' + err + ')');
	        return;
	    }
	    console.log(' *** Download queue saved to queue.json');
	}); 
}

// Loads (on server start) all downloaded files from the
// download folder and all pending files
function load() {
	// All waiting downloads (from previous run)
	fs.readFile(QUEUE_DB_FN, function read(err, data) {
	    if (err) {
	        console.log(' *** Can\'t read ' + QUEUE_DB_FN + ' (no saved downloads?)');
	        return;
	    }

	    // Get all URLs for waiting downloads
    	var waiting = JSON.parse(data);
    	for (var i = 0; i < waiting.length; i++) {
    		console.log(' *** Restoring `' + waiting[i] + '`');
    		createDownload(waiting[i]);
    	}

    	if (waiting.length > 0) {
    		console.log(' *** ' + waiting.length + ' waiting downloads restored; ' +
    					' resuming download ...');
    		onStart();
    	}
	});

	// Scan the downloads directory for downloaded files and 
	// add them to the view
	createDownloadDir();
	var files = fs.readdirSync(DOWNLOAD_DIR);
	for (var i in files){
	    var stat = fs.statSync(DOWNLOAD_DIR + '/' + files[i]);
	    if (stat.isDirectory()){
	        getFiles(name, files_);
	    } else {
			var entry = {
				'id': ID_COUNTER++,
				'url': undefined, 
				'downloader': undefined,
				'state': 4,
				'percent': 0,
				'speed': 0,
				'eta': 0,
				'filename': DOWNLOAD_DIR + '/' + files[i],
				'success': true,
				'size': stat.size,
				'name': files[i],
				'errmsg': ''
			};
			ALL.push(entry);
	    }
	}
}

function createDownload(url) {
	var entry = {
		'id': ID_COUNTER++,
		'url': url, 
		'downloader': new downloader.NodeDownloader(DOWNLOAD_DIR),
		/* 0 = queued , 1 = active, 2 = finished, 3 = stopped, 4 = file only */
		'state': 0,
		'percent': 0,
		'speed': 0,
		'eta': 0,
		'filename': '',
		'success': true,
		'size': 0,
		'name': '',
		'errmsg': ''
	};
	ALL.push(entry);
}

// --------------------------------------------------------------------
// HTTP server section

// Handles REST requests
function rest(args) {
	// Add new file to download
	if (args.url) {
		// Check if the given URL is valid
		if (!(args.url.indexOf("http") === 0 || args.url.indexOf("ftp") === 0)) {
			return {'state': 500}; // Unknown URL
		}

		// Check if URL is known
		if (!args.force) { // Omit test if forced
			for (var i = 0; i < ALL.length; i++) {
				if (ALL[i].url == args.url) {
					// URL already seen
					return {'state': 501};
				}
			}
		}

		createDownload(args.url);
		// Add entry and run it (maybe if there is no other running download)
		persist();
		onStart();
		return {'state': 200};
	}

	if (args.cmd && args.cmd == 'list') {
		return {'state': 200, 'all': ALL};
	}

	if (args.cmd && args.cmd == 'stop' && args.id) {
		var elem = false;
		for (var i = ALL.length - 1; i >= 0; i--) {
			if (ALL[i].id == args.id) {
				elem = ALL[i];
				break;
			}
		}

		if (elem == undefined || elem == null) {
			// FIXME: make me nicer (more detailed response code?)
			return {'state': 400};
		}

		// Perform action
		if (elem.state == 1) {
			// Stop the current download
			elem.downloader.stopDownload();
			elem.state = 3;
			elem.success = false;
			elem.percent = 0;
			elem.speed = 0;
			elem.eta = 0;
			// Reset the current element to let a new download start
			CURRENT = false;

			// Let the next download begin (if any)
			onStart();
			persist();
			return {'state': 200};
		} else if (elem.state == 0) {
			// Simply deactivate a queued download
			elem.state = 3;
			elem.success = false;
			elem.percent = 0;
			elem.speed = 0;
			elem.eta = 0;
			persist();
			return {'state': 200};
		}
	}

	if (args.cmd && args.cmd == 'dskstat') {
		var used = usedDsk();
		var total = totalDsk();
		return {
			'state': 200, 
			'used': used, 
			'total': total, 
			'percent': (used/total)*100.0,
		};
	}

	if (args.cmd && args.cmd == 'versions') {
		// Get module versions
		var info = process.versions;
		var str = '';
		for (var key in info) {
			str += key + ' v' + info[key];
		}

		return {
			'state': 200, 
			'node': 'NodeJS v' + info.node + ' (V8 v' + info.v8 + ')',
			'app': APP_NAME + ' v' + APP_VERSION
		};
	}

	if (args.cmd && args.cmd == 'clear-all') {
		if (!rmDir(DOWNLOAD_DIR)) {
			return {'state': 501};
		}
		ALL = [];
		persist();
		return {'state': 200};
	}

	if (args.cmd && args.cmd == 'user-script') {
		// Test only for state
		if (args.test) {
			return {'state': SCRIPT.getState() == 0 ? 500 : 200};
		}

		// Run the script
		if (args.run && SCRIPT.getState() == 0) {
			return {'state': SCRIPT.run() ? 200 : 400};
		}

		if (args.exit) {
			SCRIPT.clear();
		}

		return {'state': 200, 'data': SCRIPT.getOutput(), 'finished': SCRIPT.getState() == 2};
	}

	if (args.cmd && args.cmd == 'halt' && args.code) {
		if (args.code == SHUTDOWN_CODE) {
			// Trigger shutdown
			require('child_process').execSync(SHUTDOWN_CMD);
			console.log(' *** Server will shutdown now (' + new Date().toString() + ')');
			return {'state': 200};	
		}
		return {'state': 500};
	}

	return {'state': 400};
}

// Create the HTTP server
http.createServer(function (req, res) {
	// Parse the requested URL
	var parsedUrl = url.parse(req.url, true);

	// Backend REST API on the path /rest/
	if (parsedUrl.pathname.indexOf('/rest/') == 0) {
		// Get the arguments
		var args = parsedUrl.query;

		// Leth the REST API do the work ...
		var response = rest(args);

		// Return the result
		res.writeHead(200, {'Content-Type': 'application/json'});
		res.end(JSON.stringify(response));
	} else {
		
		// Parse the requested file name to get only the file
		// requested
		var file = parsedUrl.pathname;
		var index = file.lastIndexOf('/');
		if (file.length < 1 && index < 0) {
			cry(res, 500, 'Requested filename invalid.');
			return;
		}
		file = file.substring(index + 1);

		// Empty files are redirected to the index page
		if (file.length < 1) {
			file = "index.html";
		}

		// Get the file extension (for MIME type)
		index = file.lastIndexOf('.');
		var extension = file.substring(index + 1);

		// Check if file exists in the www/ dir and can be served
		if (!fs.existsSync('www/' + file)) {
			cry(res, 404);
		} else {
			try {
				var data = fs.readFileSync('www/' + file);
				res.writeHead(200, {
					'Content-Type': mimeLookup(extension),
					'Content-Length': data.length
				});
				res.end(data);
			} catch (err) {
				cry(res, 500, err);
			}
		}
	}
}).listen(SERVER_PORT, SERVER_HOST);

// Init the server and print logging to the standard output
console.log(' *** Welcome to ' + APP_NAME + ' (v ' + APP_VERSION + ')');
console.log(' *** Server running at http://' + SERVER_HOST + ':' + SERVER_PORT + '/');
console.log(' *** Press CTRL + C to quit')

// Load all previous and possibly pending downloads after
// everything is set-up
load();
